pandas/core/frame.py

```diff
@@ -8190,11 +8190,15 @@ def update(
         if not isinstance(other, DataFrame):
             other = DataFrame(other)
 
-        other = other.reindex_like(self)
+        # reindex rows, non-matching columns get skipped
+        other = other.reindex(self.index)
 
-        for col in self.columns:
+        shared_cols = self.columns.intersection(other.columns)
```

Comment:
@MarcoGorelli I know this solution kinda side-steps the original issue of null-matching (and reindex introducing an entire NA-column that doesn't need updating I think), but happy to have your thoughts on this solution.

Reply:
I think this is fine

Something that comes to mind is that it'll still error if `other` contains a column all of `nan`, e.g.:
```python
In [1]: df1 = pd.DataFrame({'a': [NaT]})

In [2]: df2 = pd.DataFrame({'a': [np.nan]})

In [3]: df1.update(df2, overwrite=False)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [3], line 1
----> 1 df1.update(df2, overwrite=False)

File ~/pandas-dev/pandas/core/frame.py:8096, in DataFrame.update(self, other, join, overwrite, filter_func, errors)
   8093 if mask.all():
   8094     continue
-> 8096 self.loc[:, col] = expressions.where(mask, this, that)

File ~/pandas-dev/pandas/core/computation/expressions.py:258, in where(cond, a, b, use_numexpr)
    246 """
    247 Evaluate the where condition cond on a and b.
    248 
   (...)
    255     Whether to try to use numexpr.
    256 """
    257 assert _where is not None
--> 258 return _where(cond, a, b) if use_numexpr else _where_standard(cond, a, b)

File ~/pandas-dev/pandas/core/computation/expressions.py:188, in _where_numexpr(cond, a, b)
    181     result = ne.evaluate(
    182         "where(cond_value, a_value, b_value)",
    183         local_dict={"cond_value": cond, "a_value": a, "b_value": b},
    184         casting="safe",
    185     )
    187 if result is None:
--> 188     result = _where_standard(cond, a, b)
    190 return result

File ~/pandas-dev/pandas/core/computation/expressions.py:172, in _where_standard(cond, a, b)
    170 def _where_standard(cond, a, b):
    171     # Caller is responsible for extracting ndarray if necessary
--> 172     return np.where(cond, a, b)

File <__array_function__ internals>:180, in where(*args, **kwargs)

TypeError: The DType <class 'numpy.dtype[datetime64]'> could not be promoted by <class 'numpy.dtype[float64]'>. This means that no common DType exists for the given inputs. For example they cannot be stored in a single array unless the dtype is `object`. The full list of DTypes is: (<class 'numpy.dtype[datetime64]'>, <class 'numpy.dtype[float64]'>)
```
but arguably that's desirable behaviour - you wouldn't want to update with a column of an incompatible dtype, regardless of whether its values were all missing or not.

And value-dependent behaviour wouldn't be great, so I like this solution more than the originally-suggested "if all nan then skip"